spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/autoconfig/GcpCloudSqlAutoConfiguration.java

```diff
@@ -127,8 +138,59 @@ public DataSource cloudSqlDataSource(CloudSqlJdbcInfoProvider cloudSqlJdbcInfoPr
 		hikariConfig.setUsername(this.gcpCloudSqlProperties.getUserName());
 		hikariConfig.setPassword(this.gcpCloudSqlProperties.getPassword());
 		// Setting this is useful to disable connection initialization. Especially in unit tests.
-		hikariConfig.setInitializationFailFast(this.gcpCloudSqlProperties.getInitFailFast());
+		hikariConfig.setInitializationFailFast(this.gcpCloudSqlProperties.isInitFailFast());
 
 		return new HikariDataSource(hikariConfig);
 	}
+
+	/**
+	 * Set credentials to be used by the Google Cloud SQL socket factory.
+	 *
+	 * <p>
+	 * The only way to pass a {@link CredentialFactory} to the socket factory is by passing a class
+	 * name through a system property. The socket factory creates an instance of
+	 * {@link CredentialFactory} using reflection without any arguments. Because of that, the
+	 * credential location needs to be stored somewhere where the class can read it without any
+	 * context.
+	 *
+	 * <p>
+	 * If user didn't specify credentials, the socket factory already does the right thing by using
+	 * the application default credentials by default. So we don't need to do anything.
+	 */
+	private void setCredentialsProperty() {
+		try {
+			// First tries the SQL configuration credential.
+			if (this.gcpCloudSqlProperties != null
+					&& this.gcpCloudSqlProperties.getCredentialsLocation() != null
+					&& Files.exists(this.gcpCloudSqlProperties.getCredentialsLocation())) {
+				System.setProperty(SqlCredentialFactory.CREDENTIAL_LOCATION_PROPERTY_NAME,
+						this.gcpCloudSqlProperties.getCredentialsLocation().toString());
+			}
+			// Then, the global credential.
+			else if (this.gcpProperties != null
+					&& this.gcpProperties.getCredentialsLocation() != null
+					&& this.gcpProperties.getCredentialsLocation().exists()) {
+				// A resource might not be in the filesystem, but the Cloud SQL credential must.
+				if (this.gcpProperties.getCredentialsLocation().getFile() != null) {
+					System.setProperty(SqlCredentialFactory.CREDENTIAL_LOCATION_PROPERTY_NAME,
```

Comment:
To avoid the return in the middle of the method, I would need to add a condition before setting `CREDENTIAL_FACTORY_PROPERTY`. Pick your poison :)

I changed the logic a bit to extract the location to its own variable (and noticed a return lacking). I don't think we can simplify the cascading conditions (e.g., by having a huge  "no creds file exists" return condition at the beginning) because of the credentials fallback. LMKWYT